onnxmltools/convert/sparkml/operator_converters/k_means.py

```diff
@@ -0,0 +1,174 @@
+            "alpha": -2.0,
+            "beta": 1.0,
+            "transB": 1,
+        }
+        container.add_node(
+            op_type="Gemm", 
+            inputs=[operator.inputs[0].full_name, centers_variable_name, input_row_squared_sum_variable_name], 
+            outputs=[gemm_output_variable_name],
+            **gemm_attrs
+        )
+        
+        # Euclidean Distance Squared = input_row_squared_sum - 2 * input * Transpose(Center) + Transpose(centers_row_squared_sum)
+        # [N x K]
+        container.add_node(
+            op_type="Add",
+            inputs=[gemm_output_variable_name, centers_row_squared_sum_variable_name],
+            outputs=[distance_output_variable_name],
+        )
+    elif op.getDistanceMeasure() == "cosine":
+        # centers_row_norm2: [1 x K]
+        centers_row_norm2_variable_name = scope.get_unique_variable_name("centers_row_norm2")
+        centers_row_norm2 = np.linalg.norm(centers, ord = 2, axis=1).flatten().astype(np.float32)
+        container.add_initializer(
+            centers_row_norm2_variable_name, 
+            onnx_proto.TensorProto.FLOAT, 
+            [1, K], 
+            centers_row_norm2
+            )
+        
+        # input_row_norm2: [N x 1]
+        input_row_norm2_variable_name = scope.get_unique_variable_name("input_row_norm2")
+        reduce_l2_attrs = {
+            "name": scope.get_unique_operator_name("input_row_norm2"),
+            "axes": [1],
+            "keepdims": 1,
+        }
+        container.add_node(
+            op_type="ReduceL2",
+            inputs=[operator.inputs[0].full_name],
+            outputs=[input_row_norm2_variable_name],
+            **reduce_l2_attrs
+        )
+
+        # input * Transpose(Center): [N x K]
+        zeros_variable_name = scope.get_unique_variable_name("zeros")
+        container.add_initializer(
+            zeros_variable_name, 
+            onnx_proto.TensorProto.FLOAT, 
+            [1, K], 
+            np.zeros([1, K]).flatten().astype(np.float32)
+        )
+        gemm_output_variable_name = scope.get_unique_variable_name("gemm_output")
+        gemm_attrs = {
+            "name": scope.get_unique_operator_name("GeMM"),
+            "alpha": 1.0,
+            "transB": 1,
+        }
+        container.add_node(
+            op_type="Gemm", 
+            inputs=[operator.inputs[0].full_name, centers_variable_name, zeros_variable_name],
+            outputs=[gemm_output_variable_name],
+            **gemm_attrs
+        )
+
+        # Cosine similarity = gemm_output / input_row_norm2 / centers_row_norm2: [N x K]
+        div_output_variable_name = scope.get_unique_variable_name("div_output")
+        container.add_node(
+            op_type="Div",
+            inputs=[gemm_output_variable_name, input_row_norm2_variable_name],
+            outputs=[div_output_variable_name],
+            op_version=7, # Setting to version 7 for broadcasting support
```

Comment:
Maybe you can remove the opset but raise an exception when the requested opset is below 7. That's usually what we do.

Reply:
Done.